### PR TITLE
dkg: refactor dkg to reconnect to connecting phase

### PR DESCRIPTION
**Pull request by dB2510**
Refactors dkg sync protocol to reconnect in connecting to peers phase.

category: refactor
ticket: #1475

### CODE REVIEW

File: dkg/dkg.go [hunk 1]
```diff
@@ -308,12 +308,16 @@ func startSyncProtocol(ctx context.Context, tcpNode host.Host, key *ecdsa.Privat
 	}
 
 	for _, client := range clients {
-		err := client.AwaitConnected(ctx)
+		err := client.IsConnected(ctx)
 		if err != nil {
 			return nil, err
 		}
 	}
```
Comment (corverroos) [inline dkg/dkg.go:315]: this isn't "once all peers are connected", rather this is "once all peers were connected once". 

Suggest:
 - `IsConnected` not blocking, but just return the current  status
 - refactor this to
   - check if all clients connected
   - if not, sleep for 100ms

File: dkg/sync/client.go [hunk 2]
```diff
@@ -35,59 +35,64 @@ import (
...
-	peer      peer.ID
+
+	// Immutable state
+	hashSig []byte
+	tcpNode host.Host
+	peer    peer.ID
 }
 
 // Run blocks while running the client-side sync protocol. It returns an error if the context is closed
 // or if an established connection is dropped. It returns nil after successful Shutdown.
```
Comment (corverroos) [inline dkg/sync/client.go:65]: godoc must the updated

File: dkg/sync/client.go [hunk 3]
```diff
@@ -199,13 +205,13 @@ func (c *Client) sendMsg(stream network.Stream, shutdown bool) (*pb.MsgSyncRespo
...
 // connect returns an opened libp2p stream/connection, it will retry if instructed.
-func (c *Client) connect(ctx context.Context, retry bool) (network.Stream, error) {
+func (c *Client) connect(ctx context.Context) (network.Stream, error) {
 	for {
 		s, err := c.tcpNode.NewStream(network.WithUseTransient(ctx, "sync"), c.peer, protocolID)
 		if ctx.Err() != nil {
 			return nil, ctx.Err()
 		} else if err != nil {
-			if retry {
+			if c.reconnect {
```
Comment (corverroos) [inline dkg/sync/client.go:214]: This retries always if reconnect is true, while the instruction state:
```
If c.reconnect=true, reconnect on either relayBroke or connBroke with retry=true.
If c.reconnect=false, only reconnect on relayBroke with retry=false.
```

**Review by corverroos — CHANGES_REQUESTED** [review]

File: dkg/dkg_test.go [hunk 4]
```diff
@@ -197,44 +248,147 @@ func testDKG(t *testing.T, def cluster.Definition, p2pKeys []*ecdsa.PrivateKey)
...
+	lock, keys, _ := cluster.NewForT(t, vals, nodes, nodes, 0)
+
+	tests := []struct {
+		name       string
+		connect    []int // Initial connections
+		disconnect []int // Drop some peers
+		reconnect  []int // Connect remaining peers
+	}{
+		{
+			name:       "first",
```
Comment (corverroos) [inline dkg/dkg_test.go:270]: not very discriptive

File: dkg/dkg_test.go [hunk 5]
```diff
@@ -197,44 +248,147 @@ func testDKG(t *testing.T, def cluster.Definition, p2pKeys []*ecdsa.PrivateKey)
...
+
+			// Start bootnode
+			bootnode, errChan := startBootnode(ctx, t)
+
+			dir, err := os.MkdirTemp("", "")
+			require.NoError(t, err)
+
+			configs := getConfigs(t, lock.Definition, keys, dir, bootnode)
+
+			dkgs := make([]*testDkg, nodes)
```
Comment (corverroos) [inline dkg/dkg_test.go:296]: canonical slice initialisation: `var dkgs []*testDkg`

File: dkg/dkg_test.go [hunk 6]
```diff
@@ -197,44 +248,147 @@ func testDKG(t *testing.T, def cluster.Definition, p2pKeys []*ecdsa.PrivateKey)
...
+
+			// Start DKG for initial peers.
+			for _, idx := range test.connect {
+				dkgs[idx] = startNewDKG(configs[idx], errChan)
+			}
+
+			// Drop some peers.
+			time.Sleep(time.Second * 5) // Give some time to connect.
```
Comment (corverroos) [inline dkg/dkg_test.go:304]: not a fan of sleeping. Since it is brittle and will flap. Rather wait for callbacks.

File: dkg/dkg_test.go [hunk 7]
```diff
@@ -197,44 +248,147 @@ func testDKG(t *testing.T, def cluster.Definition, p2pKeys []*ecdsa.PrivateKey)
...
+			for _, idx := range test.reconnect {
+				dkgs[idx] = startNewDKG(configs[idx], errChan)
+			}
+
+			// Assert DKG results
+			for i := 0; i < nodes; i++ {
+				dkgs[i].AssertDone(t)
+			}
+			verifyDKGResults(t, lock.Definition, dir)
+			require.NoError(t, os.RemoveAll(dir))
```
Comment (corverroos) [inline dkg/dkg_test.go:321]: no need to delete test dirs (you might not have perms)

File: dkg/dkg_test.go [hunk 8]
```diff
@@ -197,44 +248,147 @@ func testDKG(t *testing.T, def cluster.Definition, p2pKeys []*ecdsa.PrivateKey)
...
-	dir, err := os.MkdirTemp("", "")
-	require.NoError(t, err)
+	conf := dkg.Config{
+		DataDir: dir,
+		P2P: p2p.Config{
+			UDPBootnodes: []string{bootnode},
+		},
+		Log:     log.DefaultConfig(),
+		TestDef: &def,
+	}
```
Comment (corverroos) [inline dkg/dkg_test.go:336]: move this into the loop to avoid having to think and worry about mutable state.

File: dkg/dkg_test.go (hunk 5 above)
Comment (corverroos) [reply dkg/dkg_test.go:296]: add comment maybe?

Comment (dB2510) [reply dkg/dkg_test.go:296]: sure

File: dkg/dkg.go [hunk 9]
```diff
@@ -307,11 +308,36 @@ func startSyncProtocol(ctx context.Context, tcpNode host.Host, key *ecdsa.Privat
...
-		if err != nil {
-			return nil, err
+	// Check if all clients connected
+	for {
+		// Return if there is a context error.
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		connectedCount := 0
```
Comment (corverroos) [inline dkg/dkg.go:318]: `var count int`

**Pull request merged**